File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraysDivByK = function(nums, k) {
    // Creating a HashMap:
    // [Key, Value] = [Integer, Integer]
    const map = new Map();
    
    // Putting the first value as (0,1)
    map.set(0,1);
    
    // Variables:
    let sum = 0;
    let ans = 0;
    
    // Iterating the given array:
    for(let el of nums){
        sum += el;
        // Finding the Remainder:
        let rem = sum % k;
        
        // Checking whether the value of rem < 0
        if(rem < 0)
            rem += k;
        
        // Checking if the element is present in the HM or not
        if(map.has(rem)){
            ans += map.get(rem);
            map.set(rem, map.get(rem)+1);
        }
        else{
            map.set(rem, 1);
        }
    }
    
    return ans;
};

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        HashMap<Integer, Integer> h = new HashMap<>();
        // Putting the 1st value (0,1) => means 0 comes 1 times.
        h.put(0,1);
        
        // Variables:
        int sum = 0;
        int ans = 0;
        
        // Iterating
        for(int i : nums){
            sum += i;
            // Finding the remainder:
            int rem = sum % k;
            
            // Checking the value is less or greater then 0
            if(rem < 0) 
                rem += k;
            
            // Checking:
            if(h.containsKey(rem)){
                ans += h.get(rem);
            }
            
            // Putting the rem into the HashMap:
            h.put(rem, h.getOrDefault(rem, 0) +1);
        }
        
        return ans;
    }
}